File: Archiv/analyze_abortion_difference.m
% Analyze why MATLAB aborts at step 37 while Python continues
% Simulate the abortion check logic

fprintf('=== Analyzing Abortion Logic Difference ===\n\n');

% Simulate what would happen at step 37
% Assume we have 37 steps, and check what the abortion condition would be

% Typical scenario: if time steps are getting very small
% Let's simulate with different time step patterns

N = 10;  % Number of last steps to consider
tFinal = 2.0;

% Scenario 1: Time steps gradually decreasing (MATLAB pattern)
fprintf('Scenario 1: Gradually decreasing time steps (MATLAB pattern)\n');
% Assume steps 28-37 have very small time steps
small_steps = [1e-6, 1e-7, 1e-8, 1e-9, 1e-10, 1e-11, 1e-12, 1e-13, 1e-14, 1e-15];
lastNsteps = sum(small_steps);
% Assume we're at t = 0.1 (still far from tFinal)
currt = 0.1;
remTime = tFinal - currt;
ratio = remTime / lastNsteps;
fprintf('  lastNsteps = %.6e\n', lastNsteps);
fprintf('  remTime = %.6f\n', remTime);
fprintf('  ratio = remTime / lastNsteps = %.6e\n', ratio);
fprintf('  Would abort? %s (threshold: 1e9)\n\n', string(ratio > 1e9));

% Scenario 2: Time steps staying reasonable (Python pattern)
fprintf('Scenario 2: Reasonable time steps (Python pattern)\n');
% Assume steps maintain reasonable size
reasonable_steps = [0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01];
lastNsteps2 = sum(reasonable_steps);
ratio2 = remTime / lastNsteps2;
fprintf('  lastNsteps = %.6e\n', lastNsteps2);
fprintf('  remTime = %.6f\n', remTime);
fprintf('  ratio = remTime / lastNsteps = %.6e\n', ratio2);
fprintf('  Would abort? %s (threshold: 1e9)\n\n', string(ratio2 > 1e9));

% Scenario 3: Check Python's explicit zero check
fprintf('Scenario 3: Python explicit zero check\n');
lastNsteps3 = 0;
fprintf('  lastNsteps = %d\n', lastNsteps3);
fprintf('  Python would abort immediately (lastNsteps == 0)\n');
fprintf('  MATLAB: remTime / lastNsteps = Inf, which is > 1e9, so would also abort\n\n');

% Key insight: The difference might be in how time steps are computed
% If MATLAB's time steps become very small due to numerical issues or
% different adaptation logic, it would trigger abortion earlier

fprintf('=== Hypothesis ===\n');
fprintf('MATLAB time steps are becoming very small (possibly due to:\n');
fprintf('  1. Different numerical precision\n');
fprintf('  2. Different time step adaptation logic\n');
fprintf('  3. Different handling of edge cases in aux_optimaldeltat\n');
fprintf('  4. Different finitehorizon computation\n');
fprintf('This causes remTime / lastNsteps to exceed 1e9, triggering abortion.\n');
fprintf('Python time steps stay larger, so the ratio never exceeds 1e9.\n');

File: Archiv/analyze_abortion_difference.py
"""Analyze why Python continues while MATLAB aborts"""
import numpy as np

print("=== Analyzing Abortion Logic Difference ===\n")

# Simulate the abortion check logic
N = 10  # Number of last steps to consider
tFinal = 2.0

# Scenario 1: Time steps gradually decreasing (MATLAB pattern)
print("Scenario 1: Gradually decreasing time steps (MATLAB pattern)")
# Assume steps 28-37 have very small time steps
small_steps = np.array([1e-6, 1e-7, 1e-8, 1e-9, 1e-10, 1e-11, 1e-12, 1e-13, 1e-14, 1e-15])
lastNsteps = np.sum(small_steps)
# Assume we're at t = 0.1 (still far from tFinal)
currt = 0.1
remTime = tFinal - currt
ratio = remTime / lastNsteps
print(f"  lastNsteps = {lastNsteps:.6e}")
print(f"  remTime = {remTime:.6f}")
print(f"  ratio = remTime / lastNsteps = {ratio:.6e}")
print(f"  Would abort? {ratio > 1e9} (threshold: 1e9)\n")

# Scenario 2: Time steps staying reasonable (Python pattern)
print("Scenario 2: Reasonable time steps (Python pattern)")
# Assume steps maintain reasonable size
reasonable_steps = np.array([0.01] * 10)
lastNsteps2 = np.sum(reasonable_steps)
ratio2 = remTime / lastNsteps2
print(f"  lastNsteps = {lastNsteps2:.6e}")
print(f"  remTime = {remTime:.6f}")
print(f"  ratio = remTime / lastNsteps = {ratio2:.6e}")
print(f"  Would abort? {ratio2 > 1e9} (threshold: 1e9)\n")

# Scenario 3: Check Python's explicit zero check
print("Scenario 3: Python explicit zero check")
lastNsteps3 = 0
print(f"  lastNsteps = {lastNsteps3}")
print("  Python would abort immediately (lastNsteps == 0)")
print("  MATLAB: remTime / lastNsteps = Inf, which is > 1e9, so would also abort\n")

# Key insight: The difference might be in how time steps are computed
print("=== Hypothesis ===")
print("MATLAB time steps are becoming very small (possibly due to:")
print("  1. Different numerical precision")
print("  2. Different time step adaptation logic")
print("  3. Different handling of edge cases in aux_optimaldeltat")
print("  4. Different finitehorizon computation")
print("This causes remTime / lastNsteps to exceed 1e9, triggering abortion.")
print("Python time steps stay larger, so the ratio never exceeds 1e9.")
